### ISSUE TITLE
refactor: quantity adjuster for products is a reusable component

### ISSUE THREAD

**Issue by henrylin03** [henrylin03/le-vesinet#25]
# overview
the input box for adjusting quantity is the same across all individual product pages, and the cart (for each product in the cart).

however, a challenge I am facing currently is that on individual product pages, the state of the quantity is just a number (which gets displayed in the controlled `<input>`), and the setter for state can just accept a number which will update the quantity. contrast this with the _cart_, as I would like changes to individual products' quantities to update the cart immediately.

# relationship with other issues
this issue blocks #8, and should be done before #7 to prevent rework in the way state is managed

# fallback / workaround
a workaround is to clone the current site, where there is a "update cart" anchor, which only after clicked, would the state of the cart change. 

but, it just feels nicer to be able to make those changes immediately in the cart, navigate away, and come back and the quantities are still post-change.

**Comment by henrylin03** [comment]
# potential solutioning
this is presuming we don't have an "update cart" anchor on the cart page - if the user adjusts the quantity in the cart page, it updates the global state `cartProducts`

## `<QuantityInput />` UI component 

- does not have an internal state or state setter 
- **props** that it accepts:
  - **state**: the number displayed
  - **state setter function**: can be used for an `onBlur` and/or `onChange` handler of the input box. this is required in the cart page to update the global, `cartProducts` state (passed from App.jsx). the state setter for the individual product pages is passed in so that the state (quantity) in the input box is added to the `cartProducts` global state `onClick` of the "Add to cart" button

## instance on individual product pages

- **has its own state** for the quantity (this is so that it can be added to the cart `onClick` of "Add to cart" button)
- **props** it _accepts_:
  - **state setter function for `cartProducts` global state**: this is so that it can be called `onClick` of "Add to cart" button. we only need the method for _adding_ the items as users do not need the ability to remove items on product page (at least not in MVP)
- **props** it _passes_ to `<QuantityInput />`:
  -**state**: its own state for the quantity
  -**state setter function**: very basic, comes with the `useState` declaration. again, just to update the number that is displayed within the `<input>` 

## instance on shopping cart page
- does not have an internal state or state setter
- **props** that it accepts:
  - **state**: global state `cartProducts`, so that each item's quantity is displayed within the component (just passes it down to `<QuantityInput />` component). this is passed down from `App.jsx`, where it is managed.
  - **state setter function**: to add or remove items in the `cartProducts` global state. this is passed down from `App.jsx`, where it is managed.
- needs an event handler ???

## `App.jsx`
manages the global state `cartProducts` that tracks, in an object, the product ID (key) and count in the cart (value)

**Comment by henrylin03** [comment]
# potential solutioning (workaround)
this is presuming we have an "update cart" anchor on the cart page

## `<QuantityInput />` UI component 

- does not have an internal state or state setter 
- **props** that it accepts:
  - **state**: the number displayed
  - **state setter function**: can be used for an `onBlur` and/or `onChange` handler of the input box. this is required in the cart page to update the global, `cartProducts` state (passed from App.jsx). the state setter for the individual product pages is passed in so that the state (quantity) in the input box is added to the `cartProducts` global state `onClick` of the "Add to cart" button

## instance on individual product pages

- **has its own state** for the quantity (this is so that it can be added to the cart `onClick` of "Add to cart" button)
- **props** it _accepts_:
  - **state setter function for `cartProducts` global state**: this is so that it can be called `onClick` of "Add to cart" button. we only need the method for _adding_ the items as users do not need the ability to remove items on product page (at least not in MVP)
- **props** it _passes_ to `<QuantityInput />`:
  -**state**: its own state for the quantity
  -**state setter function**: very basic, comes with the `useState` declaration. again, just to update the number that is displayed within the `<input>` 

## instance on shopping cart page
- **has its own state** for the quantity _of each item_ (this is so that it can amend the cart (global state) `onClick` of "Update cart" anchor)
- **props** that it accepts:
  - **state**: global state `cartProducts`, so that each item's quantity is displayed within the component (just passes it down to `<QuantityInput />` component). this is passed down from `App.jsx`, where it is managed.
  - **state setter function(s)**: to add and remove items in the `cartProducts` global state. this is passed down from `App.jsx`, where it is managed.

## `App.jsx`
manages the global state `cartProducts` that tracks, in an object, the product ID (key) and count in the cart (value)

**Comment by henrylin03** [comment]
> # potential solutioning (workaround)
> this is presuming we have an "update cart" anchor on the cart page

to make it less annoying and more responsive, consider only displaying the anchor _if there's changes in the cart_ - this is in a separate issue #26

**Comment by henrylin03** [comment]
i have used the workaround for mvp, and will consider updating global state on change/blur events on cart page in #27

**Comment by henrylin03** [comment]
moving into waiting - changes were made as part of #7 and #8 tickets
